### PR TITLE
Add `storage-jpa-db` property into Quarkus. Distinguish postgres and …

### LINKED ISSUE #17305: Add `storage-jpa-db` property into Quarkus to distinguish postgres and crdb
https://github.com/keycloak/keycloak/pull/16549#issue-1550586541

### PR DESCRIPTION
**Pull request by vramik**
…crdb for jpa map store.

The PR adds new option --storage-jpa-db with two expected values: "postgres", "cockroach".

Database enum now contains an indication whether the database could be enabled on

- legacy store only
- new store only
- enabled for both.

It also sets that there is used XA driver with jpa map store.

Closes #17305

### CODE REVIEW

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/StoragePropertyMappers.java
Comment (mabartos) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/StoragePropertyMappers.java:308]: We could also map the property to some internal `spi` one as well, right?

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelpAll.unix.approved.txt
Comment (mabartos) [inline quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelpAll.unix.approved.txt:92]: Like this, right?

```suggestion
                       postgresql, cockroach. Default: postgresql.
```

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartHelpAll.unix.approved.txt
Comment (mabartos) [inline quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartHelpAll.unix.approved.txt:98]: Like this, right?

```suggestion
                       postgresql, cockroach. Default: postgresql.
```

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (mabartos) [inline testsuite/integration-arquillian/tests/base/pom.xml:897]: `postgres` or `postgresql` ?

**Review by mabartos — COMMENTED** [review]
@vramik Code in general looks good to me. I've just added a few nitpicks.

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (vramik) [reply testsuite/integration-arquillian/tests/base/pom.xml:897]: `postgres`

It is an alias, while `postgresql` is a databaseKind. We need to use the alias as CocroachDB and PostgreSQL have the same databaseKind and we need to distinguish between them.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelpAll.unix.approved.txt
Comment (vramik) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelpAll.unix.approved.txt:92]: `postgres` should IMO remain, but default value should be updated in the next commit.

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartHelpAll.unix.approved.txt
Comment (vramik) [reply quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartHelpAll.unix.approved.txt:98]: `postgres` should IMO remain, but default value should be updated in the next commit.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/StoragePropertyMappers.java
Comment (vramik) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/StoragePropertyMappers.java:308]: I'm not sure what `spi` you have in mind here, could you please elaborate on this?

File: quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java
Comment (vramik) [reply quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java:261]: The intention was to make easy use of methods `isEnabledOnLegacyStore` and `isEnabledOnNewStore` on filters. E.g.:

https://github.com/vramik/keycloak/blob/d760d18dbec0317f48bd43a8c9b6a775384e2dcd/quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java#L115

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/StoragePropertyMappers.java
Comment (mabartos) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/StoragePropertyMappers.java:308]: I mean keep the convention around mapping these properties to some with prefix `kc.spi-`.
```suggestion
                        .to("kc.spi-map-storage-jpa-db")
                        .mapFrom("storage")
```

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (mabartos) [reply testsuite/integration-arquillian/tests/base/pom.xml:897]: Thanks for the explanation. I see the help message was changed, and it makes sense to me right now.

File: quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java
Comment (mabartos) [reply quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java:261]: Ok. It's just a nitpick and should not block this PR.

**Review by mabartos — DISMISSED** [review]
Comments mentioned above are nitpicks and do not have to be included in the final form of this PR.

@vramik LGTM, thanks.

File: quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java
Comment (ahus1) [inline quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java:188]: AFAIK this should be a valid Hibernate class name. As there is no `CockroachDialect`, I'm suggesting as the most recent dialect supported in Hibernate 5.x.

At the same time, this is in the Quarkus 2.x branch not used, as we'll override it for the new store with the `JsonbPostgreSQL95Dialect`, so this explains why the tests are green.

```suggestion
                "org.hibernate.dialect.CockroachDB201Dialect",
```

**Review by ahus1 — CHANGES_REQUESTED** [review]
Looking at the code, IMHO we're not making use of this CLI option in Quarkus 2.x yet? I know we'll need it for Quarkus 3.x, and it is a good preparation for this. I'll also soon need it for the next cycle of CRDB optimizations, so I'm very happy to see the option being available.

There is one small change, besides that it's all fine.

File: quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java
Comment (vramik) [reply quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java]: Thanks, for pointing that out. You're right, the `CockroachDialect` is supported since Hibernate 6. I'll make sure it's reflected in https://github.com/keycloak/keycloak/pull/16549

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/StoragePropertyMappers.java
Comment (vramik) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/StoragePropertyMappers.java:309]: Updated

**Review by mabartos — APPROVED** [review]
@vramik LGTM!

@vmuzikar @pedroigor Do you want to check it on behalf of cloud-native-maintainers?

**Review by ahus1 — APPROVED** [review]
Thank you for this change. I'm looking forward to use this option soon on in one of my upcoming PRs.

File: quarkus/config-api/src/main/java/org/keycloak/config/StorageOptions.java
Comment (vmuzikar) [inline quarkus/config-api/src/main/java/org/keycloak/config/StorageOptions.java:343]: I have a bit silly question. Why do we even need this new option? :) Can't we simply piggyback on `--db`?

I understand that this is a new Store option. But what will happen with the `--db` after the new Store replaces the legacy Store? Would we keep using `--storage-jpa-db`? Isn't it a bit long? :)

File: quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java
Comment (vmuzikar) [reply quarkus/config-api/src/main/java/org/keycloak/config/database/Database.java:261]: @vramik I believe even with an `Enum` this would be achievable, if I'm not mistaken, so @mabartos has a valid point. But it's just a nitpick, non-blocking from my side.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/TransactionPropertyMappers.java
Comment (vmuzikar) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/TransactionPropertyMappers.java:43]: Just out of curiosity, why to make this change?

File: quarkus/config-api/src/main/java/org/keycloak/config/StorageOptions.java
Comment (vramik) [reply quarkus/config-api/src/main/java/org/keycloak/config/StorageOptions.java:343]: At this moment there is a difference in "expected values' as well as in "default value" between `--db` and `--storage-jpa-db`. 

That's the reason it was introduced in the first place. 

I agree that after the new Store replaces the legacy Store it might be a bit long :) 

note: https://github.com/keycloak/keycloak/issues/13562

Comment (ahus1) [reply quarkus/config-api/src/main/java/org/keycloak/config/StorageOptions.java:343]: The two options have a different set of possible values. All storage options are considered preview anyway, so I'm ok with having an extra option for now. Once we have per-realm-configuration of storages, this will soon look very different anyway.

Comment (vmuzikar) [reply quarkus/config-api/src/main/java/org/keycloak/config/StorageOptions.java:343]: Makes sense.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/TransactionPropertyMappers.java
Comment (vmuzikar) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/TransactionPropertyMappers.java:43]: No problem, let's keep it here.

**Review by vmuzikar — APPROVED** [review]
LGTM, thanks @vramik.

Do we need an approval from a store maintainer?

**Review by hmlnarik — APPROVED** [review]
Approving based on @ahus1 's approval

**Pull request merged**
